Client/Application/AppContainer.swift

```diff
@@ -0,0 +1,75 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+/// TODO: Preparing services to be injected (https://mozilla-hub.atlassian.net/browse/FXIOS-4381) & actually
+/// injecting at those points (https://mozilla-hub.atlassian.net/browse/FXIOS-4152) will come in a later PR.
+
+import Foundation
+import os.log
+import Dip
+import Storage
+
+/// This is our concrete dependency container. It holds all dependencies / services the app would need through
+/// a session.
+class AppContainer: ServiceProvider {
+
+    static let shared: ServiceProvider = AppContainer()
+
+    /// The item holding registered services.
+    private var container: DependencyContainer?
+
+    init() {
+        container = bootstrapContainer()
+    }
+
+    /// Any services needed by the client can be resolved by calling this and passing the type.
+    func resolve(type: Any.Type) -> Any {
+        do {
+            return try container?.resolve(type.self) as Any
+        } catch {
+            /// If a service we've thought was registered but can't be resolved, this is likely an issue within
+            /// bootstrapping. Double check your registrations and their types.
+            os_log(.error, "Could not resolve the requested type!")
+
+            /// We've made bad assumptions, and there's something very wrong with container setup! This is fatal.
+            fatalError("\(error)")
+        }
+    }
+
+    // MARK: - Misc helpers
+
+    /// Prepares the container by registering all services for the app session.
+    /// - Returns: A bootstrapped `DependencyContainer`.
+    private func bootstrapContainer() -> DependencyContainer {
+        return DependencyContainer { container in
+            do {
+                unowned let container = container
+
+                container.register(.eagerSingleton) {
+                    BrowserProfile(
+                        localName: "profile",
+                        syncDelegate: UIApplication.shared.syncDelegate) as Profile
+                }
+
+                /// TabManager can remain a singleton until we support multiple scenes.
+                container.register(.singleton) {
+                    TabManager(
+                        profile: try container.resolve(),
+                        imageStore: DiskImageStore(
+                            files: (try container.resolve() as Profile).files,
+                            namespace: "TabManagerScreenshots",
+                            quality: UIConstants.ScreenshotQuality))
+                }
```

Comment:
Since TabManager is a global state object it's going to be difficult to instantiate it here because then everywhere will need to point to this reference. For now I would have this return the TabManager that's been instantiated in the app delegate or update the one in the app delegate to be a computed property that resolves the value from the container. Either way they will both need to point to the same object.
Not a blocker since this isn't used anywhere yet, just something to keep in mind.

Reply:
Yup, the plan's to update the one in AppDelegate to resolve from the container. And eventually have all clients grab what they need from the container :)